src/runtime_functions.js

```diff
@@ -94,9 +94,30 @@ function convertJsFunctionToWasm(func, sig) {
 
 var freeTableIndexes = [];
 
+#if WASM_BACKEND
+// Weak map of functions in the table to their indexes, created on first use.
+var functionsInTableMap;
+#endif
+
 // Add a wasm function to the table.
 function addFunctionWasm(func, sig) {
   var table = wasmTable;
+
+#if WASM_BACKEND
+  // Check if the function is already in the table, to ensure each function
+  // gets a unique index. First, create the map if this is the first use.
+  if (!functionsInTableMap) {
+    functionsInTableMap = new WeakMap();
+    for (var i = 0; i < table.length; i++) {
+      functionsInTableMap[table.get(i)] = i;
+    }
+  }
+  if (functionsInTableMap[func]) {
+    return functionsInTableMap[func];
+  }
```

Comment:
One downside of doing this is that it will mask the current bug we are trying to investigate with @awtcode...

Reply:
If this fixes a bug, then it's a proper fix for that bug, isn't it? That is, this isn't doing a random thing that might fix a function pointer comparison - it is literally making function pointers properly unique, which *should* fix bugs.

Maybe I don't see what you mean here?

Reply:
The problem is that the module in which the bug occurs seem to be confused about how to find the address of a certain function.

For internal (dll-local) functions the compiler just puts the function in the elem section and should never call the `fp$` function.   For external (non-dll-local) function the compiler should *not* put the function in the elem section and only call the `fp$` function.

This fix will allow such confused modules to silently work.

Reply:
Hmm, I don't see that as silent. I think this PR changes the model entirely to what should have been the proper model all along: the Table has unique elements. That's more efficient and guaranteed to be correct.

As a bonus, it means we don't need to worry about coordinating things like different ways to add things to the Table. You're right that this might fix a bug in coordinating such things, but it does so by allowing us to remove that code! That's essentially what the corresponding binaryen PR does.